opendevin/controller/agent_controller.py

```diff
@@ -83,106 +76,79 @@ def __init__(
+        await self.event_stream.add_event(ErrorObservation(message), EventSource.AGENT)
 
-    async def add_history(
-        self, action: Action, observation: Observation, add_to_stream=True
-    ):
-        if not isinstance(action, Action):
-            raise TypeError(
-                f'action must be an instance of Action, got {type(action).__name__} instead'
-            )
-        if not isinstance(observation, Observation):
-            raise TypeError(
-                f'observation must be an instance of Observation, got {type(observation).__name__} instead'
-            )
+    async def add_history(self, action: Action, observation: Observation):
         self.state.history.append((action, observation))
         self.state.updated_info.append((action, observation))
-        if add_to_stream:
-            await self.event_stream.add_event(action, EventSource.AGENT)
-            await self.event_stream.add_event(observation, EventSource.AGENT)
-
-    async def _run(self):
-        if self._agent_state != AgentState.RUNNING:
-            raise ValueError('Task is not in running state')
 
-        for i in range(self._cur_step, self.max_iterations):
-            self._cur_step = i
+    async def _start_step_loop(self):
+        while True:
             try:
-                finished = await self.step(i)
-                if finished:
-                    await self.set_agent_state_to(AgentState.FINISHED)
-                    break
-            except Exception:
-                logger.error('Error in loop', exc_info=True)
-                await self.set_agent_state_to(AgentState.ERROR)
-                await self.add_error_to_history(
-                    'Oops! Something went wrong while completing your task. You can check the logs for more info.'
-                )
+                await self._step()
+            except asyncio.CancelledError:
+                logger.info('AgentController task was cancelled')
                 break
-
-            if self._is_stuck():
-                logger.info('Loop detected, stopping task')
-                await self.set_agent_state_to(AgentState.ERROR)
-                await self.add_error_to_history(
-                    'I got stuck into a loop, the task has stopped.'
+            except Exception as e:
+                logger.error(f'Error while running the agent: {e}')
+                await self.report_error(
+                    'There was an unexpected error while running the agent'
                 )
+                await self.set_agent_state_to(AgentState.ERROR)
                 break
-            await asyncio.sleep(
-                0.001
-            )  # Give back control for a tick, so other async stuff can run
-        final_state = self.get_agent_state()
-        if final_state == AgentState.RUNNING:
-            await self.set_agent_state_to(AgentState.PAUSED)
+
+            await asyncio.sleep(1)
 
     async def on_event(self, event: Event):
```

Comment:
Do you think if it is feasible to also add iteration info into `Event`? And add `source` and `iteration` to the `.to_json()` method of an Event?

Reply:
`source` definitely, that should be there already--will make sure

`iteration` is tough though. That's really the responsibility of the agent controller, and any number of events could come through that don't necessarily have to do with an iteration.

We _could_ add a `step` event, which would have an iteration number. Then you could go through the event history and know roughly the ordering. Would that work?

Reply:
> a step event, which would have an iteration number. Then you could go through the event history and know roughly the ordering. Would that work?

This looks great to me! Basically some sort of dummy barrier / separator and should be easy to use!

Reply:
Or, if we are using the definition that each `turn` starts with an `Agent` action, maybe we can simply use `source=Agent` as a separator, WDYT?

Reply:
Oh that's probably much better for now. Any `Action(source=agent)` should work perfectly as a separator

Reply:
OK `source` is added now